psyneulink/core/llvm/codegen.py

```diff
@@ -260,6 +350,44 @@ def visit_Name(self, node):
 
     def visit_Attribute(self, node):
         val = self.visit(node.value)
+
+        # special case numpy attributes
+        if node.attr == "shape":
+            # recursively extract shape from value
+            dimensions = []
+            val = val.type.pointee
+            while hasattr(val, "count"):
+                dimensions.append(val.count)
+                val = val.element
+
+            # dimensions is a tuple, so it should be a llvm struct
+            dimensions_struct = self.builder.alloca(ir.LiteralStructType([self.ctx.float_ty] * len(dimensions)))
```

Comment:
this is a struct of floats to match the python return type (tuple), as tuples are interpreted as structs in the compiled environment.

this could technically be fixed by converting arrays to structs and structs to arrays upon return (to explicitly cast it to the expected types). I am not sure if this would have unintended side-effects, however.

Reply:
representing tuples as structs is just a simplification and not a requirement
Python tuples allow dynamic indexing but struct type only works with constants so using structs is more restrictive.